core/sail/base/src/main/java/org/eclipse/rdf4j/sail/base/SailSourceConnection.java

```diff
@@ -712,17 +714,32 @@ public boolean addInferredStatement(Resource subj, IRI pred, Value obj, Resource
 
 	private void add(Resource subj, IRI pred, Value obj, SailDataset dataset, SailSink sink, Resource... contexts)
 			throws SailException {
-		if (contexts.length == 0) {
-			if (hasConnectionListeners() && !hasStatement(dataset, subj, pred, obj, NULL_CTX)) {
-				notifyStatementAdded(vf.createStatement(subj, pred, obj));
+		if (contexts.length == 0 || (contexts.length == 1 && contexts[0] == null)) {
+			if (hasConnectionListeners()) {
+				if (!hasStatement(dataset, subj, pred, obj, NULL_CTX)) {
+					notifyStatementAdded(vf.createStatement(subj, pred, obj));
+					sink.approve(subj, pred, obj, null);
+				}
+			} else {
+				sink.approve(subj, pred, obj, null);
 			}
-			sink.approve(subj, pred, obj, null);
 		} else {
 			for (Resource ctx : contexts) {
-				if (hasConnectionListeners() && !hasStatement(dataset, subj, pred, obj, ctx)) {
-					notifyStatementAdded(vf.createStatement(subj, pred, obj, ctx));
+				Resource[] contextsToCheck;
+				if (contexts.length == 1) {
+					contextsToCheck = contexts;
+				} else {
+					contextsToCheck = new Resource[] { ctx };
+				}
+
+				if (hasConnectionListeners()) {
+					if (!hasStatement(dataset, subj, pred, obj, contextsToCheck)) {
+						notifyStatementAdded(vf.createStatement(subj, pred, obj, ctx));
+						sink.approve(subj, pred, obj, ctx);
+					}
+				} else {
+					sink.approve(subj, pred, obj, ctx);
 				}
-				sink.approve(subj, pred, obj, ctx);
 			}
 		}
```

Comment:
`hasStatement(s,p,o,c)` is a vararg, which means that the JVM needs to create an array for time it is called with one or more arguments. The typical case for us though is that `contexts` is either an empty array or an array with a single element.

Reply:
Of all the tests in the MemoryStore, there is only one test that calls down to this code with more than one context.